includes/ConsoleTasks/RunJobQueueTask.php

```diff
@@ -106,13 +106,19 @@ public function execute()
                 $database->rollBack();
                 try {
                     $database->beginTransaction();
-
+                    
                     /** @var JobQueue $job */
                     $job = JobQueue::getById($job->getId(), $database);
                     $job->setDatabase($database);
-                    $job->setStatus(JobQueue::STATUS_FAILED);
-                    $job->setError($ex->getMessage());
-                    $job->setAcknowledged(0);
+                    if (str_contains($ex->getMessage, 'mwoauth-invalid-authorization')) {
```

Comment:
This string appears to only come from [this exception handler](https://gerrit.wikimedia.org/g/mediawiki/extensions/OAuth/+/7b4988f1613f958938fa48f074fec546651ace2a/src/SessionProvider.php#151), but it's also a substring used by a [number of different errors](https://codesearch.wmcloud.org/search/?q=mwoauth-invalid-authorization&files=&excludeFiles=i18n&repos=Extension%3AOAuth).

I think this will retry any authorization problem, including quite a few scenarios where a retry won't help at all.

The `Nonce already used` error [appears to be hard-coded at the moment](https://gerrit.wikimedia.org/g/mediawiki/extensions/OAuth/+/7b4988f1613f958938fa48f074fec546651ace2a/src/Lib/OAuthServer.php#267) so it might be better to match on that string instead?

Reply:
I'm also concerned that this might cause a job to loop indefinitely through the job queue. I was wondering if we could use the acknowledged flag to determine if a job has been retried already, but it looks like we set it to null when transitioning to "waiting": https://github.com/enwikipedia-acc/waca/blob/997d771885c5f63aa86c209de9e661cc7a264ad8/includes/ConsoleTasks/RunJobQueueTask.php#L56-L58

I think if we can filter on just the nonce error, the risk of a loop is mitigated. I'd hope as a community we'd also notice if jobs got stuck in a loop.